src/libsystemd/sd-daemon/sd-daemon.c

```diff
@@ -551,6 +552,41 @@ _public_ int sd_pid_notify_with_fds(
         return r;
 }
 
+_public_ int sd_notify_barrier(int unset_environment, uint64_t timeout) {
+        _cleanup_close_pair_ int pipe_fd[2] = { -1, -1 };
+        int r;
+
+        if (pipe2(pipe_fd, O_CLOEXEC) < 0) {
+                r = -errno;
+                goto finish;
+        }
+
+        struct pollfd pfd = {
+                .fd = pipe_fd[0],
+                /* POLLHUP is implicit */
+                .events = 0,
+        };
+
+        r = sd_pid_notify_with_fds(0, unset_environment, "BARRIER=1", &pipe_fd[1], 1);
+        if (r <= 0)
+                goto finish;
+
+        pipe_fd[1] = safe_close(pipe_fd[1]);
+
+        /* roughly convert to milliseconds, then clamp down */
+        timeout = timeout/1000 < INT_MAX ? timeout/1000 : INT_MAX;
+
+        if (poll(&pfd, 1, (int) timeout) >= 0) {
```

Comment:
use ppoll(), so that we can pass a timespec, and convert the specified uint64_t time to the timespace via timespec_store(). Let's not needlessly lose granularity here...

Reply:
One key thing that we're missing after converting to unsigned timeout is being able to specify an infinite timeout, is that okay?

Reply:
we usually handle UINT64_MAX specially, to mean "infinity". There's even a define for that, USEC_INFINITY (which is used mostly internally though, i.e. where we use usec_t rather than uint64_t though). i.e.

use `timeout == UINT64_MAX ? NULL : timespec_store(…)` as argument